enterprise/server/crypter_service/crypter_service.go

```diff
@@ -622,3 +623,159 @@ func (c *Crypter) testGetLastCacheRefreshRun() time.Time {
+	}
+	// This is where we'd fail if the customer supplied an invalid key, so we
+	// intentionally use a different error code here.
+	encGroupKeyPart, err := groupKeyClient.Encrypt(groupKeyPart, []byte(u.GetGroupID()))
+	if err != nil {
+		return status.UnavailableErrorf("could not use customer key for encryption: %s", err)
+	}
+
+	// We're good to go. Now just need to update the database.
+
+	key := &tables.EncryptionKey{
+		EncryptionKeyID: keyID,
+		GroupID:         u.GetGroupID(),
+	}
+	keyVersion := &tables.EncryptionKeyVersion{
+		EncryptionKeyID:    keyID,
+		Version:            1,
+		MasterEncryptedKey: encMasterKeyPart,
+		GroupKeyURI:        groupKeyURI,
+		GroupEncryptedKey:  encGroupKeyPart,
+	}
+	err = c.env.GetDBHandle().Transaction(ctx, func(tx *gorm.DB) error {
+		if err := tx.Create(key).Error; err != nil {
+			return err
+		}
+		if err := tx.Create(keyVersion).Error; err != nil {
+			return err
+		}
+		if err := tx.Exec("UPDATE `Groups` SET cache_encryption_enabled = true WHERE group_id = ?", u.GetGroupID()).Error; err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return status.InternalErrorf("could not update key information: %s", err)
+	}
+	return nil
+}
+
+func (c *Crypter) disableEncryption(ctx context.Context) error {
+	u, err := c.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return err
+	}
+	err = c.env.GetDBHandle().Transaction(ctx, func(tx *gorm.DB) error {
+		q := `
+			DELETE FROM EncryptionKeyVersions 
+			WHERE encryption_key_id IN (
+				SELECT encryption_key_id
+				FROM EncryptionKeys
+				WHERE group_id = ?
+			)
+		`
+		if err := tx.Exec(q, u.GetGroupID()).Error; err != nil {
+			return err
+		}
+		if err := tx.Exec("DELETE FROM EncryptionKeys where group_id = ?", u.GetGroupID()).Error; err != nil {
+			return err
+		}
+		if err := tx.Exec("UPDATE `Groups` SET cache_encryption_enabled = false WHERE group_id = ?", u.GetGroupID()).Error; err != nil {
+			return err
+		}
+		return nil
+	})
+	return err
+}
+
+func (c *Crypter) SetEncryptionConfig(ctx context.Context, req *enpb.SetEncryptionConfigRequest) (*enpb.SetEncryptionConfigResponse, error) {
+	if c.env.GetCrypter() == nil {
+		return nil, status.FailedPreconditionErrorf("crypter service not enabled")
+	}
+	if c.env.GetKMS() == nil {
+		return nil, status.FailedPreconditionErrorf("KMS service not enabled")
+	}
+
+	u, err := c.env.GetAuthenticator().AuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	g, err := c.env.GetUserDB().GetGroupByID(ctx, u.GetGroupID())
+	if err != nil {
+		return nil, err
+	}
+	if g.CacheEncryptionEnabled == req.Enabled {
+		return &enpb.SetEncryptionConfigResponse{}, nil
+	}
+
+	if req.Enabled {
+		if err := c.enableEncryption(ctx, req.KeyUri); err != nil {
+			return nil, err
+		}
+		return &enpb.SetEncryptionConfigResponse{}, nil
+	} else {
+		if err := c.disableEncryption(ctx); err != nil {
```

Comment:
should we also clear out the key URI here? or maybe pass it through from the req

Otherwise it seems there's not a way to clear out / change the key URI without enabling encryption

Reply:
Disabling encryption deletes the keys from the database which deletes the key URI as well. 
Once encryption is enabled, it will not be possible to change the key URI without first disabling encryption.